pycalphad/io/cs_dat.py

```diff
@@ -1159,6 +1162,999 @@ def parse_cs_dat(instring):
+
+    Parameters
+    ----------
+    dbf : Database
+        A pycalphad Database.
+    fd : file-like
+        File descriptor.
+    if_incompatible : string, optional ['raise', 'warn', 'fix']
+        Strategy if the database does not conform to the most restrictive database specification.
+        The 'warn' option (default) will write out the incompatible database with a warning.
+        The 'raise' option will raise a DatabaseExportError.
+        The 'ignore' option will write out the incompatible database silently.
+        The 'fix' option will rectify the incompatibilities e.g. through name mangling.
+    """
+    # Needed for database queries
+    from tinydb import where
+    # Before writing anything, check that the TDB is valid and take the appropriate action if not
+    if if_incompatible not in ['warn', 'raise', 'ignore', 'fix']:
+        raise ValueError('Incorrect options passed to \'if_invalid\'. Valid args are \'raise\', \'warn\', or \'fix\'.')
+
+    # Begin constructing the written database
+    writetime = datetime.datetime.now()
+    maxlen = 80
+    output = ""
+    # Comment header block
+    # Import here to prevent circular imports
+    from pycalphad import __version__
+    try:
+        # getuser() will raise on Windows if it can't find a username: https://bugs.python.org/issue32731
+        username = getpass.getuser()
+    except:
+        # if we can't find a good username, just choose a default and move on
+        username = 'user'
+    output += " Date: {} ".format(writetime.strftime("%Y-%m-%d %H:%M"))
+
+    # DAT standard is elements written from highest atomic number to lowest
+    element_order = []
+    elements = []
+    for element in dbf.elements:
+        try:
+            element_order.append(atomic_number_map.index(element.capitalize())+1)
+            elements.append(element)
+        except ValueError:
+            if element.capitalize() not in ('Va','/-'):
+                print(f'Element {element} not found in element list.')
+                print(f'{element} will be ignored and database write will attempt to continue.')
+
+    element_order = np.argsort(element_order)[::-1]
+    elements_ordered = [elements[i] for i in element_order]
+    n_elements = len(elements_ordered)
+    output += "System: {} ".format('-'.join([elements[i] for i in element_order]))
+    output += "| Generated by {} (pycalphad {})\n".format(username, __version__)
+
+    # Make a list of phase types (models) to support for writing DAT
+    supported_phase_types = ['IDMX','QKTO','RKMP','SUBL','RKMPM','SUBLM','SUBG','SUBQ'] # TODO: 'SUBI'
+
+    # Get numbers of solution phases (and type/species for each) and pure condensed phases
+    solution_phases = []
+    stoichiometric_phases = []
+    solution_phase_types = []
+    solution_phase_species = []
+
+    # DAT *always* includes ideal gas phase (gas_ideal) in header (in first solution phase position), even if not used
+    solution_phases.insert(0,'GAS_IDEAL')
```

Comment:
Certainly works with an ideal gas in. Frankly I don't yet understand other gas models well enough yet, I should probably read up on the options in both the TDB and DAT formats before deciding how to handle.

In my experience with other TDB -> DAT converters, they often break on pressure-dependent terms. This might be a tricky spot.